File: python/meshly/utils/json_schema.py
"""
Pydantic models for JSON Schema validation.

Provides type-safe JSON Schema parsing and validation.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Union

from pydantic import BaseModel, Field


class JsonSchemaProperty(BaseModel):
    """A single property in a JSON schema."""
    
    type: Union[str, None] = None
    """The type of the property (string, integer, number, boolean, array, object, null, or custom like vertex_buffer)."""
    
    title: str | None = None
    """Human-readable title."""
    
    description: str | None = None
    """Human-readable description."""
    
    default: Any = None
    """Default value."""
    
    ref: str | None = Field(default=None, alias="$ref")
    """Reference to a definition."""
    
    anyOf: list[JsonSchemaProperty] | None = None
    """Union types (e.g., Optional)."""
    
    allOf: list[JsonSchemaProperty] | None = None
    """Intersection types."""
    
    oneOf: list[JsonSchemaProperty] | None = None
    """Exclusive union types."""
    
    # Array-specific
    items: JsonSchemaProperty | None = None
    """Schema for array items."""
    
    minItems: int | None = None
    maxItems: int | None = None
    
    # Object-specific
    properties: dict[str, JsonSchemaProperty] | None = None
    """Properties for object types."""
    
    additionalProperties: Union[JsonSchemaProperty, bool, None] = None
    """Schema for additional properties (dict pattern)."""
    
    required: Union[List[str], None] = None
    """Required property names."""
    
    # String-specific
    pattern: Union[str, None] = None
    minLength: Union[int, None] = None
    maxLength: Union[int, None] = None
    format: Union[str, None] = None
    enum: Union[List[Any], None] = None
    
    # Number-specific
    minimum: Union[float, None] = None
    maximum: Union[float, None] = None
    exclusiveMinimum: Union[float, None] = None
    exclusiveMaximum: Union[float, None] = None
    multipleOf: Union[float, None] = None
    
    # Const
    const: Any = None
    
    model_config = {"extra": "allow", "populate_by_name": True}
    
    def is_array_type(self) -> bool:
        """Check if this is a meshly array type (not a JSON Schema array like list[str])."""
        # vertex_buffer and index_sequence are always meshly types
        if self.type in {"vertex_buffer", "index_sequence"}:
            return True
        # type="array" with items is a JSON Schema list, without items is a meshly array
        if self.type == "array":
            return self.items is None
        return False
    
    def is_resource_type(self) -> bool:
        """Check if this is a resource type (Resource)."""
        # Direct resource type or Resource schema
        return self.type == "resource" or self.title == "Resource"
    
    def is_optional(self) -> bool:
        """Check if this property is optional (anyOf with null)."""
        if self.anyOf:
            return any(opt.type == "null" for opt in self.anyOf)
        return False
    
    def get_inner_type(self) -> Union[JsonSchemaProperty, None]:
        """Get the non-null type from an Optional (anyOf with null)."""
        if self.anyOf:
            non_null = [opt for opt in self.anyOf if opt.type != "null"]
            return non_null[0] if non_null else None
        return self


class JsonSchema(BaseModel):
    """
    A validated JSON Schema document.
    
    This model validates that a JSON schema dict has the expected structure
    for use with DynamicModelBuilder.
    
    Usage:
        schema_dict = json.loads(schema_json)
        schema = JsonSchema.model_validate(schema_dict)
        # Now schema is validated and type-safe
    """
    
    schema_uri: Union[str, None] = Field(default=None, alias="$schema")
    """JSON Schema version URI."""
    
    id: Union[str, None] = Field(default=None, alias="$id")
    """Schema identifier."""
    
    title: Union[str, None] = None
    """Human-readable title for the schema."""
    
    description: Union[str, None] = None
    """Human-readable description."""
    
    type: Union[Literal["object"], None] = None
    """Root type (should be 'object' for Pydantic models)."""
    
    properties: Dict[str, JsonSchemaProperty] = Field(default_factory=dict)
    """Property definitions."""
    
    required: List[str] = Field(default_factory=list)
    """Required property names."""
    
    defs: Dict[str, JsonSchemaProperty] = Field(default_factory=dict, alias="$defs")
    """Reusable definitions."""
    
    additionalProperties: Union[bool, JsonSchemaProperty, None] = None
    """Whether additional properties are allowed."""
    
    model_config = {"extra": "allow", "populate_by_name": True}
    
    def get_property(self, name: str) -> Union[JsonSchemaProperty, None]:
        """Get a property by name."""
        return self.properties.get(name)
    
    def is_required(self, name: str) -> bool:
        """Check if a property is required."""
        return name in self.required
    
    def resolve_ref(self, ref: str) -> Union[JsonSchemaProperty, None]:
        """Resolve a $ref to its definition."""
        if ref.startswith("#/$defs/"):
            def_name = ref[8:]  # Remove "#/$defs/"
            return self.defs.get(def_name)
        return None
    
    def get_resolved_property(self, name: str) -> Union[JsonSchemaProperty, None]:
        """Get a property, resolving $ref if present."""
        prop = self.get_property(name)
        if prop and prop.ref:
            return self.resolve_ref(prop.ref)
        return prop
    
    def get_encoding(self, field_name: str) -> str:
        """Get the encoding type for a field (array, vertex_buffer, index_sequence)."""
        prop = self.get_resolved_property(field_name)
        if not prop:
            return "array"
        
        # Check direct type
        if prop.type and prop.is_array_type():
            return prop.type
        
        # Check anyOf (Optional types)
        if prop.anyOf:
            for opt in prop.anyOf:
                if opt.is_array_type():
                    return opt.type
                # Check $ref in anyOf
                if opt.ref:
                    resolved = self.resolve_ref(opt.ref)
                    if resolved and resolved.is_array_type():
                        return resolved.type
        
        return "array"
    
    def field_names(self) -> List[str]:
        """Get all non-metadata field names."""
        return [name for name in self.properties.keys() if not name.startswith("$")]

File: typescript/src/json-schema.ts
/**
 * TypeScript interfaces for JSON Schema validation.
 * Matches Python's json_schema.py for cross-platform compatibility.
 */

/**
 * Array encoding types for serialization.
 * - "array": Generic meshoptimizer vertex buffer (works for any array)
 * - "vertex_buffer": Optimized for 3D vertex data (N x components)
 * - "index_sequence": Optimized for mesh indices (1D array)
 */
export type ArrayEncoding = "array" | "vertex_buffer" | "index_sequence"

/**
 * A single property in a JSON schema.
 */
export interface JsonSchemaProperty {
    /** The type of the property */
    type?: string

    /** Human-readable title */
    title?: string

    /** Human-readable description */
    description?: string

    /** Default value */
    default?: unknown

    /** Reference to a definition ($ref) */
    $ref?: string

    /** Union types (e.g., Optional) */
    anyOf?: JsonSchemaProperty[]

    /** Intersection types */
    allOf?: JsonSchemaProperty[]

    /** Exclusive union types */
    oneOf?: JsonSchemaProperty[]

    /** Schema for array items */
    items?: JsonSchemaProperty

    /** Min items for array */
    minItems?: number

    /** Max items for array */
    maxItems?: number

    /** Properties for object types */
    properties?: Record<string, JsonSchemaProperty>

    /** Schema for additional properties (dict pattern) */
    additionalProperties?: JsonSchemaProperty | boolean

    /** Required property names */
    required?: string[]

    /** String pattern */
    pattern?: string

    /** Min string length */
    minLength?: number

    /** Max string length */
    maxLength?: number

    /** String format */
    format?: string

    /** Enum values */
    enum?: unknown[]

    /** Numeric minimum */
    minimum?: number

    /** Numeric maximum */
    maximum?: number

    /** Exclusive minimum */
    exclusiveMinimum?: number

    /** Exclusive maximum */
    exclusiveMaximum?: number

    /** Multiple of */
    multipleOf?: number

    /** Const value */
    const?: unknown
}

/**
 * A validated JSON Schema document.
 */
export interface JsonSchema {
    /** JSON Schema version URI */
    $schema?: string

    /** Schema identifier */
    $id?: string

    /** Human-readable title for the schema */
    title?: string

    /** Human-readable description */
    description?: string

    /** Root type (should be 'object' for Pydantic models) */
    type?: "object" | string

    /** Property definitions */
    properties?: Record<string, JsonSchemaProperty>

    /** Required property names */
    required?: string[]

    /** Reusable definitions */
    $defs?: Record<string, JsonSchemaProperty>

    /** Whether additional properties are allowed */
    additionalProperties?: boolean | JsonSchemaProperty
}

/**
 * Utility functions for working with JSON Schema.
 */
export class JsonSchemaUtils {
    /**
     * Check if a property is a meshly array type (not a JSON Schema array like list[str]).
     */
    static isArrayType(prop: JsonSchemaProperty): boolean {
        // vertex_buffer and index_sequence are always meshly types
        if (prop.type === "vertex_buffer" || prop.type === "index_sequence") {
            return true
        }
        // type="array" with items is a JSON Schema list, without items is a meshly array
        if (prop.type === "array") {
            return prop.items === undefined
        }
        return false
    }

    /**
     * Check if a property is a resource type.
     */
    static isResourceType(prop: JsonSchemaProperty): boolean {
        return prop.type === "resource" || prop.title === "Resource"
    }

    /**
     * Check if a property is optional (anyOf with null).
     */
    static isOptional(prop: JsonSchemaProperty): boolean {
        if (prop.anyOf) {
            return prop.anyOf.some(opt => opt.type === "null")
        }
        return false
    }

    /**
     * Get the non-null type from an Optional (anyOf with null).
     */
    static getInnerType(prop: JsonSchemaProperty): JsonSchemaProperty | undefined {
        if (prop.anyOf) {
            const nonNull = prop.anyOf.filter(opt => opt.type !== "null")
            return nonNull.length > 0 ? nonNull[0] : undefined
        }
        return prop
    }

    /**
     * Resolve a $ref to its definition in the schema.
     */
    static resolveRef(schema: JsonSchema, ref: string): JsonSchemaProperty | undefined {
        if (ref.startsWith("#/$defs/")) {
            const defName = ref.slice(8) // Remove "#/$defs/"
            return schema.$defs?.[defName]
        }
        return undefined
    }

    /**
     * Get a property, resolving $ref if present.
     */
    static getResolvedProperty(
        schema: JsonSchema,
        name: string
    ): JsonSchemaProperty | undefined {
        const prop = schema.properties?.[name]
        if (prop?.$ref) {
            return JsonSchemaUtils.resolveRef(schema, prop.$ref)
        }
        return prop
    }

    /**
     * Get the encoding type for a field (array, vertex_buffer, index_sequence).
     */
    static getEncoding(schema: JsonSchema, fieldName: string): ArrayEncoding {
        const prop = JsonSchemaUtils.getResolvedProperty(schema, fieldName)
        if (!prop) {
            return "array"
        }

        // Check direct type
        if (prop.type && JsonSchemaUtils.isArrayType(prop)) {
            return prop.type as ArrayEncoding
        }

        // Check anyOf (Optional types)
        if (prop.anyOf) {
            for (const opt of prop.anyOf) {
                if (JsonSchemaUtils.isArrayType(opt)) {
                    return opt.type as ArrayEncoding
                }
                // Check $ref in anyOf
                if (opt.$ref) {
                    const resolved = JsonSchemaUtils.resolveRef(schema, opt.$ref)
                    if (resolved && JsonSchemaUtils.isArrayType(resolved)) {
                        return resolved.type as ArrayEncoding
                    }
                }
            }
        }

        return "array"
    }

    /**
     * Get all non-metadata field names.
     */
    static fieldNames(schema: JsonSchema): string[] {
        return Object.keys(schema.properties ?? {}).filter(name => !name.startsWith("$"))
    }
}
